## `try ... catch`

- 能力有限
  - 对于异步任务无法捕获
- 代码侵入性强
  - 需要在函数中使用 `try catch` 对可能出现的代码块进行包裹

## `window.onerror`

- 代码侵入性下
- 对语法错误和网络错误无能为力
- 对异步和非异步错误都能捕获到运行时错误
- 如果想使用 `window.onerror` 函数消化错误，则需要显式的返回 true，以保证错误不会向上抛，控制台也不会出现一堆错误信息。

```javascript
window.onerror = function (message, source, lineno, colno, error) {
  //
}

// message: 错误提示信息
// source: 错误脚本地址
// lineno: 错误代码所在的行号
// colno: 错误代码所在的列号
// error: 错误对象信息
```
